File: src/binary_reader.cpp
//  (C) Copyright Johan Ditmar, Karel Hruda, Paul O'Neill & Luke Stedman 2009.
//  Use, modification and distribution are subject to the Boost Software License,
//  Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
//  http://www.boost.org/LICENSE_1_0.txt).

#if defined(_MSC_VER)
#    pragma warning(disable:4503 4996)
#endif

#include <protean/binary_reader.hpp>
#include <protean/variant.hpp>
#include <protean/object_proxy.hpp>
#include <protean/object_factory.hpp>
#include <protean/variant_base.hpp>
#include <protean/detail/data_table.hpp>
#include <protean/detail/data_table_column_serializers.hpp>

#include <boost/foreach.hpp>
#include <boost/scoped_ptr.hpp>
#include <boost/scoped_array.hpp>

namespace protean {
    
    binary_reader::binary_reader(std::istream &is, int mode) :
        m_is(is),
        m_mode(mode),
        m_factory(NULL)
    {
    }

    binary_reader::~binary_reader()
    {
        close();
    }

    void binary_reader::read_value(variant::enum_type_t type, variant& value)
    {
        switch( type )
        {
            case variant::Double:
            {
                double v;
                read( v ); value = v;
                break;
            }
            case variant::String:
            {
                std::string s;
                read( s ); value = s;
                break;
            }
            case variant::Int32:
            {
                boost::int32_t i;
                read( i ); value = i;
                break;
            }
            case variant::UInt32:
            {
                boost::uint32_t i;
                read( i ); value = i;
                break;
            }
            case variant::Boolean:
            {
                bool b;
                read( b ); value = b;
                break;
            }
            case variant::Int64:
            {
                boost::int64_t i;
                read( i ); value = i;
                break;
            }
            case variant::UInt64:
            {
                boost::uint64_t i;
                read( i ); value = i;
                break;
            }
            case variant::Float:
            {
                float v;
                read( v ); value = v;
                break;
            }
            case variant::Date:
            {
                variant::date_t v;
                read( v );
                
                value = v;

                break;
            }
            case variant::Time:
            {
                variant::time_t t;
                read( t ); value = t;
                break;
            }
            case variant::DateTime:
            {
                variant::date_time_t dt;
                read( dt ); value = dt;
                break;
            }
            case variant::Dictionary:
            case variant::Bag:
            {
                value = variant( static_cast<variant::enum_type_t>(type) );
                boost::uint32_t size; read( size );

                for (boost::uint32_t i=0; i<size; ++i)
                {
                    std::string name;
                    read(name);

                    variant &val = value.insert(name, variant(), variant::ReturnItem);
                    read( val );
                }
                break;
            }
            case variant::List:
            case variant::Tuple:
            {
                boost::uint32_t size;
                read(size);
                value = variant(static_cast<variant::enum_type_t>(type), size);

                for (boost::uint32_t i=0; i<size; ++i)
                {
                    read(value[i]);
                }
                break;
            }
            case variant::Buffer:
            {
                boost::uint32_t size;
                read(size);

                value = variant(variant::Buffer, size);
                void* data = (void*)value.as<void*>();
                read(data, size);

                break;
            }
            case variant::TimeSeries:
            {
                value = variant(static_cast<variant::enum_type_t>(type));
                boost::uint32_t size; read(size);

                for (boost::uint32_t i=0; i<size; ++i)
                {
                    variant::date_time_t date_time;
                    read(date_time);

                    variant &val = value.push_back(date_time, variant(), variant::ReturnItem);
                    read(val);
                }
                break;
            }
            case variant::DataTable:
            {
                boost::int32_t numRows, numCols;
                read( numCols );
                read( numRows );

                value = variant( variant::DataTable, numRows );

				{
					std::vector<boost::int32_t> colTypes( numCols );
					BOOST_FOREACH( boost::int32_t& colType, colTypes )
						read( colType );

					std::vector<std::string> colNames( numCols );
					BOOST_FOREACH( std::string& colName, colNames )
						read( colName );

					for ( boost::int32_t i( 0 ); i != numCols; ++i )
						value.add_column( static_cast<variant::enum_type_t>( colTypes[i] ), colNames[i] );
				}

                detail::data_table& dt = value.m_value.get<variant::DataTable>();
                BOOST_FOREACH(detail::data_table::column_container_type::reference column, dt.columns())
                {
                    // Default-allocate `rows' column values to be read into (no reallocation required since
                    // capacity of `rows' was specified in DataTable construction)
                    column.resize( numRows );

                    if (column.type() & variant_base::Primitive)
                    {
                        boost::scoped_ptr<detail::data_table_column_reader> column_reader(
                            detail::make_data_table_column_binary_reader(column, *this)
                        );

                        while (column_reader->has_next())
                        {
                            column_reader->read();
                            column_reader->advance();
                        }
                    }
                    else
                    {
                        variant::iterator iter(column.begin());
                        variant::iterator end(column.end());
                        while (iter != end)
                            read(*(iter++));
                    }
                }

                break;
            }
            case variant::Object:
            {
                std::string class_name;
                boost::int32_t version;

                read(class_name);
                read(version);

                handle<object> obj;
                if (m_factory!=NULL)
                {
                    obj = m_factory->create_instance(class_name);
                    if (obj.null())
                    {
                        if ((m_mode & binary_mode::CreateProxy)!=0)
                        {
                            boost::throw_exception(variant_error("Unable to create object from factory"));
                        }
                    }
                }

                if (obj.null())
                {
                    obj = new object_proxy(class_name);
                }

                // read parameter dictionary
                variant params;
                read(params);
                obj->inflate(params, version);

                value = variant(obj);
        
                break;
            }
            case variant::Exception:
            {
                std::string type, message, source, stack;
                read(type);
                read(message);
                read(source);
                read(stack);
                value = exception_data(type, message, source, stack);
                break;
            }
            case variant::None:
                break;
            case variant::Any:
            {
                std::string s;
                read(s);
                value = variant(s);
                value.m_type = variant::Any;
                break;
            }
            case variant::Array:
            {
                boost::uint32_t size;
                read(size);
                boost::uint32_t array_type;
                read(array_type);

                handle<typed_array> a(new typed_array(size, static_cast<variant::enum_type_t>(array_type)));
                for (boost::uint32_t i=0; i<size; ++i)
                {
                    variant v;
                    read_value(static_cast<variant::enum_type_t>(array_type), v);
                    a->at(i).swap(v);
                }
                value = a;
                break;
            }
        default:        
            boost::throw_exception(variant_error("Case exhaustion: " + variant::enum_to_string(type)));        
        }
    }

    void binary_reader::read(variant& value)
    {
        boost::uint32_t type = variant::None;
        read(type);
        read_value(static_cast<variant::enum_type_t>(type), value);
    }
    void binary_reader::read(std::string& value)
    {
        boost::uint32_t length;
        read(length);

        boost::scoped_array<char> buffer( new char[length] );
        read_bytes(buffer.get(), length);
        value = std::string(buffer.get(), length);

    }
    void binary_reader::read(detail::string& value)
    {
        boost::uint32_t length;
        read(length);

        boost::scoped_array<char> buffer( new char[length] );
        read_bytes(buffer.get(), length);
        value.initialise(buffer.get(), length);
    }
    void binary_reader::read(bool& value)
    {
        boost::int32_t b;
        read(b);
        value = (b != 0);
    }
    void binary_reader::read(boost::int32_t& value)
    {
        read_bytes(reinterpret_cast<char*>(&value), 4 );
    }
    void binary_reader::read(boost::uint32_t& value )
    {
        read_bytes(reinterpret_cast<char*>(&value), 4 );
    }
    void binary_reader::read(boost::int64_t& value )
    {
        read_bytes(reinterpret_cast<char*>(&value), 8 );
    }
    void binary_reader::read(boost::uint64_t& value )
    {
        read_bytes(reinterpret_cast<char*>(&value), 8 );
    }
    void binary_reader::read(float& value)
    {
        read_bytes(reinterpret_cast<char*>(&value), sizeof(float) );
    }
    void binary_reader::read(double& value)
    {
        read_bytes(reinterpret_cast<char*>(&value), sizeof(double) );
    }
    void binary_reader::read(variant::date_t& value)
    {
        if ((m_writer_mode & binary_mode::DateTimeAsTicks)!=0)
        {
            boost::uint32_t total_days;
            read(total_days);

            value = variant::min_date() + boost::gregorian::date_duration(total_days);
        }
        else
        {
            binary_date_t date;
            read_bytes(reinterpret_cast<char*>(&date), sizeof(binary_date_t));

            value = variant::date_t( date.year, date.month, date.day );
        }
    }
    void binary_reader::read(variant::time_t& value)
    {
        if ((m_writer_mode & binary_mode::DateTimeAsTicks)!=0)
        {
            boost::int64_t total_millis;
            read(total_millis);

            // Bit nasty! Can't simply do "value = boost::posix_time::millisec(total_millis);" because of:
            //  https://svn.boost.org/trac/boost/ticket/3471
            // Hopefully this will be fixed for v1.44.
            boost::uint32_t hours = total_millis / 3600000;
            boost::uint32_t tmp = total_millis % 3600000;
            boost::uint32_t mins = tmp / 60000;
            tmp %= 60000;
            boost::uint32_t secs = tmp / 1000;
            boost::uint32_t millis = tmp % 1000;

            value = boost::posix_time::time_duration(hours, mins, secs) + boost::posix_time::millisec(millis);
        }
        else
        {
            binary_time_t time;
            read_bytes(reinterpret_cast<char*>(&time), sizeof(binary_time_t));

            value = variant::time_t(time.hour, time.minute, time.second);
        }
    }
    void binary_reader::read(variant::date_time_t& value)
    {
        if ((m_writer_mode & binary_mode::DateTimeAsTicks)!=0)
        {
            variant::time_t duration;
            read(duration);

            value = variant::min_date_time() + duration;
        }
        else
        {
            variant::date_t date;
            variant::time_t time;

            read(date);
            read(time);

            value = variant::date_time_t(date, time);
        }
    }
    void binary_reader::read(void*& data, size_t length)
    {
        // 'data' will be owned by the variant
        read_bytes(reinterpret_cast<char*>(data), length);
    }

    void binary_reader::set_factory(object_factory& factory)
    {
        m_factory = &factory;
    }

    void binary_reader::read_bytes(char* buffer, size_t length)
    {
        if (!m_filter.read(buffer, length))
        {
            boost::throw_exception(variant_error("Error reading from stream"));
        }

        // read packing bytes (if necessary)
        size_t residual = (4 - (length % 4)) % 4;
        if (residual != 0)
        {
            static int packing = 0;
            if (!m_filter.read(reinterpret_cast<char*>(&packing), residual))
            {
                boost::throw_exception(variant_error("Error reading from stream"));
            }
        }
    }

    void binary_reader::setup()
    {
        if( !m_is.good() )
        {
            boost::throw_exception(variant_error("Input stream is bad"));
        }

        // read header
        int header[3];
        if (!m_is.read(reinterpret_cast<char*>(header), sizeof(header)))
        {
            boost::throw_exception(variant_error("Error reading header from stream"));
        }

        // check magic number
        if (header[0]!=static_cast<int> (binary_magic_number))
        {
            boost::throw_exception(variant_error("Bad magic number, this looks like invalid binary data"));
        }

        // store version information
        m_major_version = (header[1] >> 16) & 0x0000FFFF;
        m_minor_version = header[1] & 0x0000FFFF;

        if (m_major_version>binary_major_version)
        {
            boost::throw_exception(variant_error((boost::format("Version of received binary data is not compatible, received version %d > this version %d")
                % static_cast<int>(m_major_version)
                % static_cast<int>(binary_major_version)
            ).str()));
        }

        m_writer_mode = header[2];

        // create compression filter if necessary
        if ((m_writer_mode & binary_mode::Compress)!=0)
        {
            bool zlib_no_header = (m_writer_mode & binary_mode::ZlibHeader) == 0;
            m_filter.push(boost::iostreams::zlib_decompressor(binary_compression_params(zlib_no_header)));
        }
        m_filter.push(m_is);
    }

    void binary_reader::close()
    {
        m_filter.reset();
    }

	bool binary_reader::eof() const
	{
		return m_filter.eof();
	}

    binary_reader& operator>>(binary_reader& reader, variant& v)
    {
        reader.setup();
        reader.read( v );
        reader.close();

        return reader;
    }

} // namespace protean

File: bindings/cs/BinaryReader.cs
//  (C) Copyright Johan Ditmar, Karel Hruda, Paul O'Neill & Luke Stedman 2009.
//  Use, modification and distribution are subject to the Boost Software License,
//  Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
//  http://www.boost.org/LICENSE_1_0.txt).

using System;
using System.Collections.Generic;
using System.Text;

using System.Data;

namespace protean {

    public class BinaryReader
    {
        public BinaryReader(System.IO.Stream stream, BinaryMode mode, IVariantObjectFactory factory)
        {
            m_mode = mode;
            m_factory = factory;
            m_stream = stream;
        }

        public BinaryReader(System.IO.Stream stream) :
            this(stream, BinaryMode.Default, null)
        { }
    
        public Variant Read()
        {
            ReadHeader();
            return ReadVariant();
        }

        public static Variant FromBytes(byte[] bytes, BinaryMode mode, IVariantObjectFactory factory)
        {
            using (System.IO.MemoryStream ms = new System.IO.MemoryStream(bytes))
            {
                BinaryReader reader = new BinaryReader(ms, mode, factory);
                return reader.Read();
            }
        }

        public static Variant FromBytes(byte[] bytes)
        {
            return FromBytes(bytes, BinaryMode.Default, null);
        }

        private void ReadHeader()
        {
            byte[] bytes = new byte[4];
            m_stream.Read(bytes, 0, 4);

            UInt32 magicNumber = System.BitConverter.ToUInt32(bytes, 0);
            if (magicNumber != BinaryConstants.PROTEAN_MAGIC)
            {
                throw new VariantException("Invalid magic number");
            }

            // ignore version info for now
            m_stream.Read(bytes, 0, 4);

            m_stream.Read(bytes, 0, 4);
            m_mode = (BinaryMode)System.BitConverter.ToUInt32(bytes, 0);

            if ((m_mode & BinaryMode.Compress) != 0)
            {
                if ((m_mode & BinaryMode.ZlibHeader) != 0)
                {
                    throw new VariantException("Binary data appears to contain ZLIB header which is currently not supported in the protean.NET BinaryReader");
                }

                m_filter = new System.IO.Compression.DeflateStream(m_stream, System.IO.Compression.CompressionMode.Decompress, true);
            }
            else
            {
                m_filter = m_stream;
            }
        }

        delegate object ReadDelegate();

        private Variant ReadVariant()
        {
            Variant.EnumType type = (Variant.EnumType)ReadUInt32();

            switch (type)
            {
                case Variant.EnumType.None:
                    return new Variant(Variant.EnumType.None);
                case Variant.EnumType.String:
                    return new Variant(ReadString());
                case Variant.EnumType.Any:
                    return new Variant(Variant.EnumType.Any, ReadString());
                case Variant.EnumType.Float:
                    return new Variant(ReadFloat());
                case Variant.EnumType.Double:
                    return new Variant(ReadDouble());
                case Variant.EnumType.Int32:
                    return new Variant(ReadInt32());
                case Variant.EnumType.UInt32:
                    return new Variant(ReadUInt32());
                case Variant.EnumType.Int64:
                    return new Variant(ReadInt64());
                case Variant.EnumType.UInt64:
                    return new Variant(ReadUInt64());
                case Variant.EnumType.Boolean:
                    return new Variant(ReadBoolean());
                case Variant.EnumType.Time:
                    if ((m_mode & BinaryMode.DateTimeAsTicks) == 0)
                    {
                        throw new VariantException("Binary data has DateTimeAsTicks mode disabled which is not supported in the protean.NET BinaryReader");
                    }
                    return new Variant(ReadTime());
                case Variant.EnumType.DateTime:
                    if ((m_mode & BinaryMode.DateTimeAsTicks) == 0)
                    {
                        throw new VariantException("Binary data has DateTimeAsTicks mode disabled which is not supported in the protean.NET BinaryReader");
                    }
                    return new Variant(ReadDateTime());
                case Variant.EnumType.Date:
                    throw new VariantException("Attempt to read Date variant which is no longer supported");
                case Variant.EnumType.Tuple:
                {
                    int length = ReadInt32();

                    Variant result = new Variant(Variant.EnumType.Tuple, length);
                    for (int i = 0; i < length; ++i)
                    {
                        result[i] = ReadVariant();
                    }
                    return result;
                }
                case Variant.EnumType.List:
                {
                    int length = ReadInt32();

                    Variant result = new Variant(type, length);
                    for (int i=0; i<length; ++i)
                    {
                        result.Add(ReadVariant());
                    }
                    return result;
                }
                case Variant.EnumType.Dictionary:
                case Variant.EnumType.Bag:
                {
                    int length = ReadInt32();

                    Variant result = new Variant(type);
                    for (int i=0; i<length; ++i)
                    {
                        String key = ReadString();
                        Variant value = ReadVariant();

                        result.Add(key, value);
                    }
                    return result;
                }
                case Variant.EnumType.TimeSeries:
                {
                    Variant result = new Variant(type);

                    int length = ReadInt32();

                    for (int i=0; i<length; ++i)
                    {
                        DateTime time = ReadDateTime();
                        Variant value = ReadVariant();

                        result.Add(time, value);
                    }

                    return result;
                }
                case Variant.EnumType.Object:
                {
                    string className = ReadString();
                    int version = ReadInt32();
                    Variant param = ReadVariant();


                    IVariantObject obj = null;
                    if (m_factory != null)
                    {
                        obj = m_factory.Create(className);

                        if (obj == null && (m_mode & BinaryMode.CreateProxy) == 0)
                        {
                            throw new VariantException("Object of class " + className + " is not regsistered in factory");
                        }
                    }

                    if (obj == null)
                    {
                        obj = new VariantObjectProxy(className);
                    }

                    obj.Inflate(param, version);

                    return new Variant(obj);
                }
                case Variant.EnumType.Exception:
                {
                    string xtype = ReadString();
                    string xmessage = ReadString();
                    string xsource = ReadString();
                    string xstack = ReadString();

                    return new Variant(new VariantExceptionInfo(xtype, xmessage, xsource, xstack));
                }
                case Variant.EnumType.Buffer:
                {
                    int length = ReadInt32();
                    return new Variant(ReadBytes(length, true));
                }
                case Variant.EnumType.DataTable:
                {
                     return new Variant(ReadDataTable());
                }
                default:
                    throw new VariantException("Case exhaustion: " + type.ToString());
            }
        }

        private byte[] ReadBytes(int length, bool readPadding)
        {
            byte[] bytes = ReadBytes(length);

            if (readPadding)
            {
                int residual = (4 - (length % 4)) % 4;
                for (int i = 0; i < residual; ++i)
                {
                    m_filter.ReadByte();
                }
            }

            return bytes;
        }
        private byte[] ReadBytes(int length)
        {
            byte[] bytes = new byte[length];
            m_filter.Read(bytes, 0, length);
            return bytes;
        }
        private bool ReadBoolean()
        {
            return ReadInt32()!=0;
        }

        private string ReadString()
        {
            Int32 length = ReadInt32();
            byte[] bytes = ReadBytes(length, true);

            return System.Text.Encoding.ASCII.GetString(bytes, 0, length);
        }
        private float ReadFloat()
        {
            return System.BitConverter.ToSingle(ReadBytes(sizeof(float)), 0);
        }
        private double ReadDouble()
        {
            return System.BitConverter.Int64BitsToDouble(ReadInt64());
        }
        private Int32 ReadInt32()
        {
            return System.BitConverter.ToInt32(ReadBytes(sizeof(Int32)), 0);
        }
        private UInt32 ReadUInt32()
        {
            return System.BitConverter.ToUInt32(ReadBytes(sizeof(UInt32)), 0);
        }
        private Int64 ReadInt64()
        {
            return System.BitConverter.ToInt64(ReadBytes(sizeof(Int64)), 0);
        }
        private UInt64 ReadUInt64()
        {
            return System.BitConverter.ToUInt64(ReadBytes(sizeof(UInt64)), 0);
        }
        private TimeSpan ReadTime()
        {
            Int64 total_millis = ReadInt64();

            return new TimeSpan(total_millis * 10000);
        }
        private DateTime ReadDateTime()
        {
            Int64 total_millis = ReadInt64();

            return Variant.MinDateTime + new TimeSpan(total_millis * 10000);
        }
        private DataTable ReadDataTable()
        {
            int numCols = ReadInt32();
            int numRows = ReadInt32();
                    
            ReadDelegate[] colReaders = new ReadDelegate[numCols];
            Variant.EnumType[] colTypes = new Variant.EnumType[numCols];

            for (int i = 0; i < numCols; ++i)
            {
                colTypes[i] = (Variant.EnumType)ReadInt32();

                switch (colTypes[i])
                {
                    case VariantBase.EnumType.Float:
                        colReaders[i] = delegate() { return ReadFloat(); };
                        break;
                    case VariantBase.EnumType.Double:
                        colReaders[i] = delegate() { return ReadDouble(); };
                        break;
                    case VariantBase.EnumType.Boolean:
                        colReaders[i] = delegate() { return ReadBoolean(); };
                        break;
                    case VariantBase.EnumType.String:
                        colReaders[i] = delegate() { return ReadString(); };
                        break;
                    case VariantBase.EnumType.Int32:
                        colReaders[i] = delegate() { return ReadInt32(); };
                        break;
                    case VariantBase.EnumType.UInt32:
                        colReaders[i] = delegate() { return ReadUInt32(); };
                        break;
                    case VariantBase.EnumType.Int64:
                        colReaders[i] = delegate() { return ReadInt64(); };
                        break;
                    case VariantBase.EnumType.UInt64:
                        colReaders[i] = delegate() { return ReadUInt64(); };
                        break;
                    case VariantBase.EnumType.Time:
                        colReaders[i] = delegate() { return ReadTime(); };
                        break;
                    case VariantBase.EnumType.DateTime:
                        colReaders[i] = delegate() { return ReadDateTime(); };
                        break;
                }
            }

            string[] colNames = new string[numCols];
            for (int i = 0; i < numCols; ++i)
            {
                colNames[i] = ReadString();
            }

            DataTable dt = new DataTable();
            for (int i = 0; i < numCols; ++i)
            {
                dt.Columns.Add(new DataColumn(colNames[i], VariantPrimitiveBase.EnumToType(colTypes[i])));
            };

            // Write number of rows
            for (int i=0; i<numRows; ++i)
            {
                DataRow dr = dt.NewRow();
                for (int j = 0; j < numCols; ++j)
                {
                    dr[j] = colReaders[j]();
                }
                dt.Rows.Add(dr);
            }

            return dt;
        }

        private System.IO.Stream m_stream;
        private System.IO.Stream m_filter;
        private BinaryMode m_mode;
        private IVariantObjectFactory m_factory;
    }

} // protean
